pageserver/src/tenant/manifest.rs

```diff
@@ -0,0 +1,252 @@
+//!    startup speed.
+//! 2. ensure all operations are atomic and avoid corruption, solving issues
+//!    like redundant image layer and preparing us for future compaction
+//!    strategies.
+//!
+//! There is also a format for storing all layer files on S3, called
+//! `index_part.json`. Compared with index_part, MANIFEST is an WAL which
+//! records all operations as logs, and therefore we can easily replay the
+//! operations when recovering from crash, while ensuring those operations
+//! are atomic upon restart.
+//!
+//! Currently, this is not used in the system. Future refactors will ensure
+//! the storage state will be recorded in this file, and the system can be
+//! recovered from this file. This is tracked in
+//! https://github.com/neondatabase/neon/issues/4418
+
+use std::io::{Read, Write};
+
+use crate::virtual_file::VirtualFile;
+use anyhow::{bail, Result};
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use crc32c::crc32c;
+use serde::{Deserialize, Serialize};
+use tracing::log::warn;
+use utils::lsn::Lsn;
+
+use super::storage_layer::PersistentLayerDesc;
+
+pub struct Manifest {
+    file: VirtualFile,
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct Snapshot {
+    pub layers: Vec<PersistentLayerDesc>,
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub enum Record {
+    AddLayer(PersistentLayerDesc),
+    RemoveLayer(PersistentLayerDesc),
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct ManifestHeader {
+    version: usize,
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub enum Operation {
+    /// The header of the manifest (always the first record)
+    Header(ManifestHeader),
+    /// A snapshot of the current state
+    Snapshot(Snapshot, Lsn),
+    /// An atomic operation that changes the state
+    Operation(Vec<Record>, Lsn),
+}
+
+struct Header {
+    size: u32,
+    checksum: u32,
+}
+
+const HEADER_LEN: usize = 8;
+
+impl Header {
+    fn encode(&self) -> BytesMut {
+        let mut buf = BytesMut::with_capacity(HEADER_LEN);
+        buf.put_u32(self.size);
+        buf.put_u32(self.checksum);
+        buf
+    }
+
+    fn decode(mut buf: &[u8]) -> Self {
+        assert!(buf.len() == HEADER_LEN, "invalid header");
+        Self {
+            size: buf.get_u32(),
+            checksum: buf.get_u32(),
+        }
+    }
+}
+
+impl Manifest {
+    pub fn init(file: VirtualFile, snapshot: Snapshot, lsn: Lsn) -> Result<Self> {
+        let mut manifest = Self { file };
+        manifest.append_operation(Operation::Header(ManifestHeader { version: 1 }))?;
+        manifest.append_operation(Operation::Snapshot(snapshot, lsn))?;
+        Ok(manifest)
+    }
+
+    /// Load a manifest. Returns the manifest and a list of operations. If the manifest is corrupted,
+    /// the bool flag will be set to true and the user is responsible to reconstruct a new manifest and
+    /// backup the current one.
```

Comment:
Why do we still return bool?

Reply:
There are two cases when load:

1. The file is *definitely* corrupted, for example, the operation record cannot be decoded, the manifest header cannot be found, etc. In this case, we return an error.
2. The file is *partially written*, for example, only the last record is corrupted due to force shutdown. In this case, we can return the full operation sequence to the user while tell the caller that the manifest is partially corrupted, and they need to generate a new one.

If we simply return an error for case (2), the caller cannot know the operation sequence. An alternative might be including the recovered sequences in the error type, do you think this would be a better implementation?

Reply:
I see. That makes sense.

But it's confusing with the current docstring. People need to fully read the code to understand when we return `corrupted = true` and when we return `Err`.

Reply:
Also it's easy to forget to check if "corrupted = true" was returned. Rust offers a lot of tools to force the caller deal with all situations. Should we upgrade this return value to an enum, or something equivalent?

Reply:
We can add `#[must_use]` so that the compiler will give a warning if the user is not checking that flag.

Reply:
... added a commit